File: api/dm.py
# from api.helpers import Snowflake
# from pymongo import ReturnDocument
# from datetime import datetime
import cherrypy
from pymongo.database import Database

from api.base import Base


@cherrypy.expose
class DirectMessages(Base):
    def __init__(self, db: Database):
        self.users = db.users
        self.direct_messages = db.direct_messages

    @cherrypy.tools.json_out()
    @cherrypy.popargs('dm_id')
    def GET(self, dm_id):
        # Read the token.
        token = cherrypy.request.headers['Token']
        # Validate the token.
        user: dict = self.users.find_one({'token': token})
        if user is None:
            raise cherrypy.HTTPError(401, 'Unauthorized')
        # Check if user is in the server.
        elif not user.__contains__(
            'direct_messages'
        ) or not user['direct_messages'].__contains__(dm_id):
            raise cherrypy.HTTPError(403, 'You cannot access this guild!')
        # Check if the DM exists.
        direct_message = self.direct_messages.find_one({'id': dm_id})
        if direct_message is None:
            raise cherrypy.HTTPError(410, 'DM channel was deleted!')
        # Send the DM.
        direct_message.pop('_id')
        return direct_message

File: src/api/dm.ts
import { Endpoint } from '../types'

export const apiDmGET: Endpoint = (db) => async (req, res) => {
  // Verify if the user exists.
  const token = req.headers['token']
  const user = await db.collection('users').findOne({ token })
  if (!user) {
    res.sendStatus(401)
    return
  }
  // Check if the user is in the DMs.
  const id = req.params['dm_id']
  if (!user['direct_messages'] || !user['direct_messages'].includes(id)) {
    res.sendStatus(403)
    return
  }
  // Get the DM.
  const directMessage = await db.collection('direct_messages').findOne({ id })
  if (!directMessage) {
    res.status(410).send('DM channel was deleted!')
    return
  }
  // Send it.
  directMessage['_id'] = undefined
  res.send(directMessage)
}
